content_handler/framebuffer_skia.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/content_handler/framebuffer_skia.h"
+
+#include <magenta/syscalls.h>
+
+#include <utility>
+
+#include "lib/ftl/logging.h"
+
+namespace flutter_content_handler {
+
+FramebufferSkia::FramebufferSkia() {}
+
+FramebufferSkia::~FramebufferSkia() {}
+
+void FramebufferSkia::Bind(mojo::InterfaceHandle<mojo::Framebuffer> framebuffer,
+                           mojo::FramebufferInfoPtr info) {
+  if (!framebuffer) {
+    FTL_LOG(ERROR) << "Failed to bind framebuffer";
+    surface_ = nullptr;
+    return;
+  }
+
+  framebuffer_.Bind(std::move(framebuffer));
+  info_ = std::move(info);
+
+  uintptr_t buffer = 0;
+  size_t row_bytes = info_->row_bytes;
+  size_t size = row_bytes * info_->size->height;
+
+  mx_status_t status =
+      mx_process_vm_map(0, info_->vmo.get().value(), 0, size, &buffer,
+                        MX_VM_FLAG_PERM_READ | MX_VM_FLAG_PERM_WRITE);
+
+  if (status < 0) {
+    FTL_LOG(ERROR) << "Cannot map framebuffer (status=" << status << ")";
+    return;
+  }
+
+  SkColorType sk_color_type;
+  switch (info_->format) {
+    case mojo::FramebufferFormat::RGB_565:
+      sk_color_type = kRGB_565_SkColorType;
+      break;
+    case mojo::FramebufferFormat::ARGB_8888:
+      sk_color_type = kRGBA_8888_SkColorType;
+      break;
+    default:
+      FTL_LOG(WARNING) << "Unknown color type " << info_->format;
+      sk_color_type = kRGB_565_SkColorType;
+      break;
+  }
+
+  SkImageInfo image_info =
+      SkImageInfo::Make(info_->size->width, info_->size->height, sk_color_type,
+                        kPremul_SkAlphaType);
+
+  surface_ = SkSurface::MakeRasterDirect(
```

Comment:
I don't see a balancing `mx_process_vm_unmap` to get rid of the mapping. It could be that the implicit `mx_handle_close` that happens when `info_` is collected in the destructor clears all existing mappings. I don't know for sure though. I think it would be better to clear the mapping ourselves in the destructor or us `MakeRasterDirectReleaseProc` and unmap it in the callback.